src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionWalker.java

```diff
@@ -0,0 +1,250 @@
+
+        // Since we're processing regions rather than individual reads, tell the progress
+        // meter to check the time more frequently (every 10 regions instead of every 1000 regions).
+        progressMeter.setRecordsBetweenTimeChecks(10L);
+
+        for ( final ReadShard readShard : readShards ) {
+            // Since reads in each shard are lazily fetched, we need to pass the filter to the window
+            // instead of filtering the reads directly here
+            readShard.setReadFilter(countedFilter);
+            currentReadShard = readShard;
+
+            processReadShard(readShard,
+                    new ReferenceContext(reference, readShard.getPaddedInterval()), // use the fully-padded window to fetch overlapping data
+                    new FeatureContext(features, readShard.getPaddedInterval()));
+        }
+
+        logger.info(countedFilter.getSummaryLine());
+    }
+
+    private void processReadShard( ReadShard shard, ReferenceContext referenceContext, FeatureContext featureContext ) {
+        // Divide each shard into one or more assembly regions using our AssemblyRegionEvaluator:
+        final Iterable<AssemblyRegion> assemblyRegions = AssemblyRegion.createFromReadShard(shard,
+                getHeaderForReads(), referenceContext, featureContext, assemblyRegionEvaluator(),
+                minAssemblyRegionSize, maxAssemblyRegionSize, assemblyRegionPadding, activeProbThreshold,
+                maxProbPropagationDistance);
+
+        // Call into the tool implementation to process each assembly region from this shard.
+        for ( final AssemblyRegion assemblyRegion : assemblyRegions ) {
+            logger.debug("Processing assembly region at " + assemblyRegion.getSpan() + " isActive: " + assemblyRegion.isActive() + " numReads: " + assemblyRegion.getReads().size() + " in read shard " + shard.getInterval());
+
+            apply(assemblyRegion,
+                    new ReferenceContext(reference, assemblyRegion.getExtendedSpan()),
+                    new FeatureContext(features, assemblyRegion.getExtendedSpan()));
+
+            progressMeter.update(assemblyRegion.getSpan());
+        }
+    }
+
+    /**
+     * Shutdown data sources.
+     *
+     * Marked final so that tool authors don't override it. Tool authors should override onTraversalDone() instead.
+     */
+    @Override
+    protected final void onShutdown() {
+        // Overridden only to make final so that concrete tool implementations don't override
+        super.onShutdown();
+    }
+
+    /**
+     * @return The evaluator to be used to determine whether each locus is active or not. Must be implemented by tool authors.
+     */
+    public abstract AssemblyRegionEvaluator assemblyRegionEvaluator();
+
+    /**
+     * Process an individual AssemblyRegion. Must be implemented by tool authors. Each region will be marked as
+     * either "active" or "inactive" using the results of the configured {@link #assemblyRegionEvaluator}.
```

Comment:
The regions are pre-marked -- re-wrote the comments here to be clearer and more precise.